### PR TITLE
573: Calculate elements for file payment

### PR DESCRIPTION
**Pull request by jorgesanchezperez**
- Added support to calculate element for file payment consent
- Clean up some classes
Issue: https://github.com/securebankingaccesstoolkit/securebankingaccesstoolkit/issues/573

### CODE REVIEW

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java [hunk 1]
```diff
@@ -16,8 +16,25 @@
...
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public abstract class PaymentConsentValidation {
+
+    protected List<OBError1> errors = new ArrayList<>();
+
+    private static final BigDecimal ZERO = new BigDecimal(0);
```
Comment (dbadham-fr) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java:37]: just an FYI... this constant exists: java.math.BigDecimal.ZERO

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java [hunk 2]
```diff
@@ -31,4 +48,153 @@ public abstract class PaymentConsentValidation extends GenericValidations {
...
+
+        if (isNull(exchangeRateInformation.getRateType(), "RateType")) {
+            return;
+        }
+        OBExchangeRateType2Code rateType = exchangeRateInformation.getRateType();
+        switch (rateType) {
+            case AGREED -> {
+                if (!exchangeRateInformation.getUnitCurrency().equals(currencyOfTransfer)) {
+                    errors.add(OBRIErrorType.DATA_INVALID_REQUEST.toOBError1(
+                            String.format("The currency of transfer should be the same with the unit currency.")
```
Comment (dbadham-fr) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java:101]: Remove String.format

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java [hunk 3]
```diff
@@ -31,4 +48,153 @@ public abstract class PaymentConsentValidation extends GenericValidations {
...
+        }
+        validateCurrency(exchangeRateInformation.getUnitCurrency());
+
+        if (isNull(exchangeRateInformation.getRateType(), "RateType")) {
+            return;
+        }
+        OBExchangeRateType2Code rateType = exchangeRateInformation.getRateType();
+        switch (rateType) {
+            case AGREED -> {
+                if (!exchangeRateInformation.getUnitCurrency().equals(currencyOfTransfer)) {
```
Comment (dbadham-fr) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java:99]: Is this validation correct? 
From the spec:
> For an "Agreed" RateType a requested exchange rate is populated in the ExchangeRate field, against the UnitCurrency. I.e, if the UnitCurrency is GBP and CurrencyOfTransfer is USD, then ExchangeRate will be 1.34 (USD to 1 GBP).

https://openbankinguk.github.io/read-write-api-site3/v3.1.10/resources-and-data-models/pisp/international-payment-consents.html#notes

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java [hunk 4]
```diff
@@ -16,8 +16,25 @@
...
+import uk.org.openbanking.datamodel.payment.OBWriteInternational2DataInitiationExchangeRateInformation;
+import uk.org.openbanking.datamodel.payment.OBWriteInternational3DataInitiationExchangeRateInformation;
 
-public abstract class PaymentConsentValidation extends GenericValidations {
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public abstract class PaymentConsentValidation {
```
Comment (dbadham-fr) [inline securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java:33]: Can you add some unit tests for these validation methods please?

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java (hunk 3 above)
Comment (jorgesanchezperez) [reply securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java]: Validation changed: `must specify ExchangeRate and ContractIdentification when requesting an Agreed RateType`.

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java (hunk 4 above)
Comment (jorgesanchezperez) [reply securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java:35]: New tests added for all payment validations.

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java (hunk 2 above)
Comment (jorgesanchezperez) [reply securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java]: Removed!

File: securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java (hunk 1 above)
Comment (jorgesanchezperez) [reply securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/validation/PaymentConsentValidation.java]: Deleted constant to use the existing one.

**Review by dbadham-fr — APPROVED** [review]

**Pull request merged**